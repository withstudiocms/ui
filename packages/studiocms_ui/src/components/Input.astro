---
import type { HTMLAttributes } from 'astro/types';
import { generateID } from '../utils/generateID';

interface Props extends HTMLAttributes<'input'> {
	label?: string;
	type?: 'text' | 'password' | 'email' | 'number' | 'tel' | 'url';
	placeholder?: string;
	isRequired?: boolean;
	name?: string;
	disabled?: boolean;
	defaultValue?: string;
	class?: string;
};

const {
	label,
	placeholder,
	name = generateID('input'),
	type = 'text',
	defaultValue,
	isRequired = false,
	disabled = false,
	class: className,
	...props
} = Astro.props;
---

<label for={name} class="sui-input-label" class:list={[disabled && "disabled"]}>
  {label && (
		<span class="label">
			{label} <span class="req-star">{isRequired && "*"}</span>
		</span>
	)}
  <input
    placeholder={placeholder}
    name={name}
    id={name}
    type={type}
    class="sui-input"
		class:list={[className]}
    required={isRequired}
    disabled={disabled}
    value={defaultValue}
		{...props}
  />
</label>
<style>
	.sui-input-label {
		width: 100%;
		display: flex;
		flex-direction: column;
		gap: .25rem;
	}

	.sui-input-label.disabled {
		opacity: 0.5;
		pointer-events: none;
		color: hsl(var(--text-muted));
	}

	.label {
		font-size: 14px;
	}

	.sui-input {
		padding: .5rem 1rem;
		border-radius: 8px;
		border: 1px solid hsl(var(--border));
		background: hsl(var(--background-step-2));
		color: hsl(var(--text-normal));
		transition: all .15s ease;
	}

	.sui-input:hover {
		background: hsl(var(--background-step-3));
	}

	.sui-input:active,
	.sui-input:focus {
		border: 1px solid hsl(var(--primary-base));
		outline: none;
		background: hsl(var(--background-step-2));
	}

	.disabled .sui-input:active {
		border: 1px solid hsl(var(--border));
	}

	.req-star {
		color: hsl(var(--danger-base));
		font-weight: 700;
	}
</style>
