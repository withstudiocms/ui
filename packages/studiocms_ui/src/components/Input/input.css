.sui-input-label {
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: .25rem;
}

.sui-input-label.disabled {
  opacity: 0.5;
  pointer-events: none;
  color: hsl(var(--text-muted));
}

.label {
  font-size: 14px;
}

.sui-input {
  padding: .5rem 1rem;
  border-radius: var(--radius-md);
  border: 1px solid hsl(var(--border));
  background: hsl(var(--background-step-2));
  color: hsl(var(--text-normal));
  transition: all .15s ease;
}

.sui-input:hover {
  background: hsl(var(--background-step-3));
}

.sui-input:active,
.sui-input:focus {
  border: 1px solid hsl(var(--primary-base));
  outline: none;
  background: hsl(var(--background-step-2));
}

.disabled .sui-input:active {
  border: 1px solid hsl(var(--border));
}

.req-star {
  color: hsl(var(--danger-base));
  font-weight: 700;
}
