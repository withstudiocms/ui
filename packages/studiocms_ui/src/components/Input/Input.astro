---
import type { HTMLAttributes } from 'astro/types';
import { generateID } from '../../utils/generateID.js';
import './input.css';

/**
 * The props for the input component.
 */
interface Props extends HTMLAttributes<'input'> {
	/**
	 * The label of the input.
	 */
	label?: string;
	/**
	 * The type of the input. Defaults to `text`.
	 */
	type?: 'text' | 'password' | 'email' | 'number' | 'tel' | 'url' | 'search';
	/**
	 * The placeholder of the input.
	 */
	placeholder?: string;
	/**
	 * Whether the input is required. Defaults to `false`.
	 */
	isRequired?: boolean;
	/**
	 * The name attribute for the input. Useful for form submission.
	 */
	name?: string;
	/**
	 * Whether the input is disabled. Defaults to `false`.
	 */
	disabled?: boolean;
	/**
	 * The default value of the input.
	 */
	defaultValue?: string;
	/**
	 * Additional classes to apply to the input.
	 */
	class?: string;
}

const {
	label,
	placeholder,
	name = generateID('input'),
	type = 'text',
	defaultValue,
	isRequired = false,
	disabled = false,
	class: className,
	...props
} = Astro.props;
---

<label for={name} class="sui-input-label" class:list={[disabled && "disabled"]}>
  {label && (
		<span class="label">
			{label} <span class="req-star">{isRequired && "*"}</span>
		</span>
	)}
  <input
    placeholder={placeholder}
    name={name}
    id={name}
    type={type}
    class="sui-input"
		class:list={[className]}
    required={isRequired}
    disabled={disabled}
    value={defaultValue}
		{...props}
  />
</label>
