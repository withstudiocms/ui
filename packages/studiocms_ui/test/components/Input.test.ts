import { describe, expect } from 'vitest';
import Input from '../../src/components/Input/Input.astro';
import { test } from '../fixtures/vitest/AstroContainer';

const mockProps = {
	name: 'input-test',
};

describe('Input Component', () => {
	test('renders Input component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Input, 'Input', {
			props: { ...mockProps },
		});
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ label: 'Input Test' },
		{ description: 'This is a test input' },
		{ placeholder: 'Enter text...' },
		{ isRequired: true },
		{ disabled: true },
		{ defaultValue: 'test-value' },
		{ icon: 'heroicons:academic-cap' },
		{ icon: { name: 'heroicons:academic-cap', position: 'left' } },
		{ icon: { name: 'heroicons:academic-cap', position: 'right' } },
	])('renders Input with "%s"', async (props, { renderComponent }) => {
		const result = await renderComponent(Input, 'Input', {
			props: { ...mockProps, ...props },
		});
		expect(result).toMatchSnapshot();
	});
});
