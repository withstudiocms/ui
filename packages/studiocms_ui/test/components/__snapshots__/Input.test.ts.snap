// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input Component > renders Input component correctly 1`] = `"<label for="input-test" class="sui-input-label">    <input name="input-test" id="input-test" type="text" class="sui-input">  </label>"`;

exports[`Input Component > renders Input with "{ defaultValue: 'test-value' }" 1`] = `"<label for="input-test" class="sui-input-label">    <input name="input-test" id="input-test" type="text" class="sui-input" value="test-value">  </label>"`;

exports[`Input Component > renders Input with "{ description: 'This is a test input' }" 1`] = `"<label for="input-test" class="sui-input-label">    <input name="input-test" id="input-test" type="text" class="sui-input"> <span class="sui-input-desc">This is a test input</span> </label>"`;

exports[`Input Component > renders Input with "{ disabled: true }" 1`] = `"<label for="input-test" class="sui-input-label disabled">    <input name="input-test" id="input-test" type="text" class="sui-input" disabled>  </label>"`;

exports[`Input Component > renders Input with "{ icon: 'heroicons:academic-cap' }" 1`] = `"<label for="input-test" class="sui-input-label">  <svg style="min-width: 20px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="input-icon icon-left" height="20" width="20" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M4.26 10.147a60 60 0 0 0-.491 6.347A48.6 48.6 0 0 1 12 20.904a48.6 48.6 0 0 1 8.232-4.41a61 61 0 0 0-.491-6.347m-15.482 0a51 51 0 0 0-2.658-.813A60 60 0 0 1 12 3.493a60 60 0 0 1 10.399 5.84q-1.345.372-2.658.814m-15.482 0A51 51 0 0 1 12 13.489a50.7 50.7 0 0 1 7.74-3.342M6.75 15a.75.75 0 1 0 0-1.5a.75.75 0 0 0 0 1.5m0 0v-3.675A55 55 0 0 1 12 8.443m-7.007 11.55A5.98 5.98 0 0 0 6.75 15.75v-1.5"/></svg>  <input name="input-test" id="input-test" type="text" class="sui-input has-icon icon-left">  </label>"`;

exports[`Input Component > renders Input with "{ icon: { name: 'heroicons:academic-cap', position: 'left' } }" 1`] = `"<label for="input-test" class="sui-input-label">   <svg style="min-width: 20px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="input-icon icon-left" height="20" width="20" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M4.26 10.147a60 60 0 0 0-.491 6.347A48.6 48.6 0 0 1 12 20.904a48.6 48.6 0 0 1 8.232-4.41a61 61 0 0 0-.491-6.347m-15.482 0a51 51 0 0 0-2.658-.813A60 60 0 0 1 12 3.493a60 60 0 0 1 10.399 5.84q-1.345.372-2.658.814m-15.482 0A51 51 0 0 1 12 13.489a50.7 50.7 0 0 1 7.74-3.342M6.75 15a.75.75 0 1 0 0-1.5a.75.75 0 0 0 0 1.5m0 0v-3.675A55 55 0 0 1 12 8.443m-7.007 11.55A5.98 5.98 0 0 0 6.75 15.75v-1.5"/></svg> <input name="input-test" id="input-test" type="text" class="sui-input has-icon icon-left">  </label>"`;

exports[`Input Component > renders Input with "{ icon: { name: 'heroicons:academic-cap', position: 'right' } }" 1`] = `"<label for="input-test" class="sui-input-label">   <svg style="min-width: 20px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="input-icon icon-right" height="20" width="20" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M4.26 10.147a60 60 0 0 0-.491 6.347A48.6 48.6 0 0 1 12 20.904a48.6 48.6 0 0 1 8.232-4.41a61 61 0 0 0-.491-6.347m-15.482 0a51 51 0 0 0-2.658-.813A60 60 0 0 1 12 3.493a60 60 0 0 1 10.399 5.84q-1.345.372-2.658.814m-15.482 0A51 51 0 0 1 12 13.489a50.7 50.7 0 0 1 7.74-3.342M6.75 15a.75.75 0 1 0 0-1.5a.75.75 0 0 0 0 1.5m0 0v-3.675A55 55 0 0 1 12 8.443m-7.007 11.55A5.98 5.98 0 0 0 6.75 15.75v-1.5"/></svg> <input name="input-test" id="input-test" type="text" class="sui-input has-icon icon-right">  </label>"`;

exports[`Input Component > renders Input with "{ isRequired: true }" 1`] = `"<label for="input-test" class="sui-input-label">    <input name="input-test" id="input-test" type="text" class="sui-input" required>  </label>"`;

exports[`Input Component > renders Input with "{ label: 'Input Test' }" 1`] = `"<label for="input-test" class="sui-input-label"> <span class="label"> Input Test <span class="req-star"></span> </span>   <input name="input-test" id="input-test" type="text" class="sui-input">  </label>"`;

exports[`Input Component > renders Input with "{ placeholder: 'Enter text...' }" 1`] = `"<label for="input-test" class="sui-input-label">    <input placeholder="Enter text..." name="input-test" id="input-test" type="text" class="sui-input">  </label>"`;
